### PR TITLE
Build target for KLEE (undone)

### PR DESCRIPTION
**Pull request by huxinwen1994**
### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### CODE REVIEW

File: cmodelres/CMakeLists.txt [hunk 1]
```diff
@@ -42,6 +42,31 @@ else()
     message(WARNING "Failed to find sea. See -DSEA_PATH.")
 endif()
 
+# Locates KLEE, as required for testing.
```
Comment (wuestholz) [inline cmodelres/CMakeLists.txt:45]: @huxinwen1994 I assume that this is just adapted from the way we locate Seahorn. Right?

File: cmodelres/CMakeLists.txt [hunk 2]
```diff
@@ -129,3 +154,15 @@ if(SEA_EXE)
         COMMAND_EXPAND_LISTS
     )
 endif()
+
+
+# target for KLEE
```
Comment (wuestholz) [inline cmodelres/CMakeLists.txt]: @huxinwen1994 Should we make this conditional like for Seahorn above?

```
if(KLEE_EXE)
...
endif()
```

File: cmodelres/CMakeLists.txt [hunk 3]
```diff
@@ -129,3 +154,15 @@ if(SEA_EXE)
         COMMAND_EXPAND_LISTS
     )
 endif()
+
+
```
Comment (wuestholz) [inline cmodelres/CMakeLists.txt:178]: We should add something like this for KLEE's options:

```
set(KLEE_ARGS "--simplify-sym-indices --write-cvcs --write-cov --output-module --max-memory=1000 --disable-inlining --optimize --use-forked-solver --use-cex-cache --libc=uclibc --posix-runtime --external-calls=all --only-output-states-covering-new --max-sym-array-size=4096 --max-instruction-time=30s --max-time=60min --watchdog --max-memory-inhibit=false --max-static-fork-pct=1 --max-static-solve-pct=1 --max-static-cpfork-pct=1 --switch-type=internal --search=random-path --search=nurs:covnew --use-batching-search --batch-instructions=10000 --silent-klee-assume --max-forks=512")
```

File: cmodelres/CMakeLists.txt [hunk 4]
```diff
@@ -129,3 +154,15 @@ if(SEA_EXE)
         COMMAND_EXPAND_LISTS
     )
 endif()
+
```
Comment (wuestholz) [inline cmodelres/CMakeLists.txt:157]: We probably want something like this:

```
set(CMAKE_C_COMPILER clang)
set(LLVM_LINK llvm-link)
set(KLEE_DRIVER_BC "klee_driver.bc")
```

File: cmodelres/CMakeLists.txt (hunk 2 above)
Comment (wuestholz) [reply cmodelres/CMakeLists.txt]: Maybe we need to define a target like for Seahorn:

```
    add_custom_target(
        symbex
	COMMAND ${KLEE_EXE} ${KLEE_ARGS} ${KLEE_DRIVER_BC}
        COMMAND_EXPAND_LISTS
    )
```

File: cmodelres/CMakeLists.txt (hunk 4 above)
Comment (wuestholz) [reply cmodelres/CMakeLists.txt:157]: We probably need a target that produces $KLEE_DRIVER_BC bitcode file:

```
    add_custom_target(
        klee_driver
	COMMAND ${CMAKE_C_COMPILER} -I ${KLEE_PATH}/include -D MC_USE_STDINT -emit-llvm -c -g -O0 -Xclang -disable-O0-optnone ${EXE_SRCS_C} libverify/verify_klee.c
    COMMAND ${LLVM_LINK} -f cmodel.bc verify_klee.bc -o ${KLEE_DRIVER_BC}
        COMMAND_EXPAND_LISTS
    )
```

File: cmodelres/CMakeLists.txt (hunk 2 above)
Comment (wuestholz) [reply cmodelres/CMakeLists.txt]: It should have klee_driver target as a dependency:

```
add_dependencies(symbex, klee_driver)
```

File: cmodelres/CMakeLists.txt (hunk 1 above)
Comment (huxinwen1994) [reply cmodelres/CMakeLists.txt:45]: yes

File: cmodelres/CMakeLists.txt [hunk 5]
```diff
@@ -129,3 +154,28 @@ if(SEA_EXE)
...
+# target for KLEE
+if(KLEE_EXE)
+    add_custom_target(
+            klee_driver
+        COMMAND ${CMAKE_C_COMPILER} -I ${KLEE_SRC} -D MC_USE_STDINT -emit-llvm -c -g -O0 -Xclang -disable-O0-optnone ${CMAKE_CURRENT_SOURCE_DIR}/cmodel.c ${CMAKE_CURRENT_SOURCE_DIR}/libverify/verify_klee.c
+        COMMAND ${LLVM_LINK} -f cmodel.bc verify_klee.bc -o ${KLEE_DRIVER_BC}
+            COMMAND_EXPAND_LISTS
+        )
+
+    set(KLEE_ARGS "--simplify-sym-indices;--write-cvcs;--write-cov;--output-module;--max-memory=1000;--disable-inlining;--optimize;--use-forked-solver;--use-cex-cache;--libc=uclibc;--posix-runtime;--external-calls=all;--only-output-states-covering-new;--max-sym-array-size=4096;--max-instruction-time=30s;--max-time=60min;--watchdog;--max-memory-inhibit=false;--max-static-fork-pct=1;--max-static-solve-pct=1;--max-static-cpfork-pct=1;--switch-type=internal;--search=random-path;--search=nurs:covnew;--use-batching-search;--batch-instructions=10000;--silent-klee-assume;--max-forks=512")
```
Comment (wuestholz) [inline cmodelres/CMakeLists.txt]: I would move this to the other declarations above (e.g., KLEE_SRC).

File: cmodelres/CMakeLists.txt [hunk 6]
```diff
@@ -129,3 +154,28 @@ if(SEA_EXE)
...
     )
 endif()
+
+set(CMAKE_C_COMPILER clang)
+set(LLVM_LINK llvm-link)
+set(KLEE_DRIVER_BC "klee_driver.bc")
+set(KLEE_SRC "/home/klee/klee_src/include")
+
+# target for KLEE
+if(KLEE_EXE)
```
Comment (wuestholz) [inline cmodelres/CMakeLists.txt:180]: Maybe this should use KLEE_PATH and we should introduce a nested if for the symbex target that checks if KLEE_EXE is set.

File: cmodelres/CMakeLists.txt [hunk 7]
```diff
@@ -129,3 +154,28 @@ if(SEA_EXE)
...
+    add_custom_target(
+            klee_driver
+        COMMAND ${CMAKE_C_COMPILER} -I ${KLEE_SRC} -D MC_USE_STDINT -emit-llvm -c -g -O0 -Xclang -disable-O0-optnone ${CMAKE_CURRENT_SOURCE_DIR}/cmodel.c ${CMAKE_CURRENT_SOURCE_DIR}/libverify/verify_klee.c
+        COMMAND ${LLVM_LINK} -f cmodel.bc verify_klee.bc -o ${KLEE_DRIVER_BC}
+            COMMAND_EXPAND_LISTS
```
Comment (wuestholz) [inline cmodelres/CMakeLists.txt]: @huxinwen1994 Maybe we need to add something like this here:

```
SOURCES ${KLEE_DEPS}
```

The list should be defined before adding the target. Something like this:

```
    set(KLEE_DEPS "")
    list(APPEND KLEE_DEPS "${CMAKE_CURRENT_SOURCE_DIR}/libverify/verify_klee.c")
    foreach(fn ${EXE_SRCS_C})
        list(APPEND KLEE_DEPS "${CMAKE_CURRENT_SOURCE_DIR}/${fn}")
    endforeach(fn)

    add_custom_target(
        klee_driver
    COMMAND ${CMAKE_C_COMPILER} -I ${KLEE_SRC} -D MC_USE_STDINT -emit-llvm -c -g -O0 -Xclang -disable-O0-optnone ${KLEE_DEPS}
        COMMAND ${LLVM_LINK} -f cmodel.bc verify_klee.bc -o ${KLEE_DRIVER_BC}
        COMMAND_EXPAND_LISTS
        SOURCES ${KLEE_DEPS}
    )
```

**Comment by wuestholz** [comment]
@huxinwen1994 BTW, I assume you checked that this works for a fresh build. You should check these two things separately and after deleting all build artifacts:
- you can build the klee-driver target only
- you can build the symbex target only

Can you please double-check this once you made the changes I described above?

**Comment by wuestholz** [comment]
@ScottWe Could you please have a look at this PR?

File: cmodelres/CMakeLists.txt (hunk 3 above)
Comment (ScottWe) [reply cmodelres/CMakeLists.txt:178]: Are these the standard args for klee? I'm unfamiliar with klee except for having set it up to test our integration.

If there's some args we don't want fixed all the time, we can take them as optional arguments, like i do with Seahorn.

File: cmodelres/CMakeLists.txt (hunk 6 above)
Comment (ScottWe) [reply cmodelres/CMakeLists.txt:180]: Yeah restructuring a bit could make sense, but given my knowledge of klee, it lgtm. My schedule's pretty packed this evening, but I'll try to propose an update, time permitting.

File: cmodelres/CMakeLists.txt (hunk 1 above)
Comment (ScottWe) [inline cmodelres/CMakeLists.txt:45]: I anticipate some conflicts with `cmodel-dev` now. We probably want to pull in that branch first.

File: cmodelres/CMakeLists.txt (hunk 3 above)
Comment (wuestholz) [reply cmodelres/CMakeLists.txt:178]: Yeah, they're mostly standard. We've been playing with the `--max-forks` a bit and we might want to make that optional in the future. I'd suggest we keep it fixed for now and add that later.

File: cmodelres/CMakeLists.txt (hunk 6 above)
Comment (wuestholz) [reply cmodelres/CMakeLists.txt:180]: Great! I think @huxinwen1994 addressed this. But if you have any other improvements please go ahead. :)

File: cmodelres/CMakeLists.txt (hunk 1 above)
Comment (wuestholz) [reply cmodelres/CMakeLists.txt:45]: Ah, just noticed that you pulled https://github.com/ScottWe/solidity-to-cmodel/pull/10 which has very slight overlap: I think we can now drop `set(CMAKE_C_COMPILER clang)` in this branch.

**Review by wuestholz — COMMENTED** [review]
@ScottWe Thanks for taking a look! Just commented on your points.

**Pull request merged**